### PR TITLE
NFT Transfers & Burns

### PR DESCRIPTION
**Pull request by redpartyhat**

### CODE REVIEW

**Comment by Lux1n** [comment]
Thank you for this, this will remove a few major pain points with the current NFT experience!

The hard work is very much appreciated ❤️

File: lib/block_view.go [hunk 1]
```diff
@@ -6290,6 +6503,11 @@ func (bav *UtxoView) _connectUpdateNFT(
 		return 0, 0, nil, RuleErrorCannotUpdateNonExistentNFT
 	}
 
+	// Verify the NFT is not a pending transfer.
+	if prevNFTEntry.IsPending {
```
Comment (diamondhands0) [inline lib/block_view.go:6507]: Do we want to do this check in _connectNFTBid() and _acceptNFTBit() as well or nah?

File: lib/block_view.go [hunk 2]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorNFTTransferRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
```
Comment (diamondhands0) [inline lib/block_view.go:7249]: s/the poster's public key/the NFT owner's public key

File: lib/block_view.go [hunk 3]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+		// public key.
+	}
+
+	// Now we are ready to transfer the NFT.
+
+	// Create the updated NFTEntry.
+	newNFTEntry := &NFTEntry{
```
Comment (diamondhands0) [inline lib/block_view.go:7256]: Hmm... this seems error-prone compared to setting only the fields that are changing. For example, if we add a new field to NFTEntry then we could forget to initialize it here. Is there a reason you did it this way vs the following?

```
// Make a copy of the previous NFT
newNFTEntry := *prevNFTEntry
// Update the fields that were set during this transfer.
newNFTEntry.OwnerPKID = receiverPKID.PKID
newNFTEntry.UnlockableText = txMeta.UnlockableTex
newNFTEntry.IsPencing = true
```

File: lib/block_view.go [hunk 4]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorAcceptNFTTransferRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
```
Comment (diamondhands0) [inline lib/block_view.go:7344]: s/poster/NFT owner

File: lib/block_view.go [hunk 5]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+		// public key.
+	}
+
+	// Now we are ready to transfer the NFT.
+
+	// Create the updated NFTEntry (everything the same except for IsPending) and set it.
+	newNFTEntry := &NFTEntry{}
```
Comment (diamondhands0) [inline lib/block_view.go:7351]: I think this works but prefer:

```
// Make a copy of the previous NFT entry
newNFTEntry := *prevNFTEntry
// Set the NFT to no longer be pending
newNFTEntry.IsPending = false
// Grab the pointer here.
bav._setNFTEntryMappings(&newNFTEntry)
```

File: lib/block_view.go (hunk 4 above)
Comment (diamondhands0) [inline lib/block_view.go:7344]: s/poster/NFT owner

File: lib/block_view.go (hunk 5 above)
Comment (diamondhands0) [inline lib/block_view.go:7351]: I think this works but prefer:

```
// Make a copy of the previous NFT entry
newNFTEntry := *prevNFTEntry
// Set the NFT to no longer be pending
newNFTEntry.IsPending = false
// Grab the pointer here.
bav._setNFTEntryMappings(&newNFTEntry)
```

Comment (diamondhands0) [inline lib/block_view.go:7351]: I think this works but prefer:

```
// Make a copy of the previous NFT entry
newNFTEntry := *prevNFTEntry
// Set the NFT to no longer be pending
newNFTEntry.IsPending = false
// Grab the pointer here.
bav._setNFTEntryMappings(&newNFTEntry)
```

File: lib/block_view.go [hunk 6]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorBurnNFTRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
```
Comment (diamondhands0) [inline lib/block_view.go:7432]: s/poster/NFT owner

File: lib/block_view.go [hunk 7]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+		// public key.
+	}
+
+	// Create a backup before we burn the NFT.
+	prevNFTEntry := &NFTEntry{}
+	*prevNFTEntry = *nftEntry
```
Comment (diamondhands0) [inline lib/block_view.go:7438]: Prefer prevNFTEntry := *nftEntry and then use &prevNFTEntry.

File: lib/block_view.go [hunk 8]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+
+	// Delete the NFT.
+	bav._deleteNFTEntryMappings(nftEntry)
+
+	// Save a copy of the previous postEntry and then increment NumNFTCopiesBurned.
+	prevPostEntry := &PostEntry{}
+	*prevPostEntry = *nftPostEntry
```
Comment (diamondhands0) [inline lib/block_view.go:7445]: Same format here. prevPostEntry := *nftPostEntry and then use &prevPostEntry

File: lib/block_view.go [hunk 9]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+	nftPostEntry.NumNFTCopiesBurned++
```
Comment (diamondhands0) [inline lib/block_view.go:7446]: Heh. Nice that this can only be incremented. Makes life easy.

File: lib/block_view_test.go [hunk 10]
```diff
@@ -18967,3 +19210,503 @@ func TestBitCloutDiamondErrorCases(t *testing.T) {
...
+		post2 := DBGetPostEntryByPostHash(db, post2Hash)
+		require.Equal(uint64(1), post2.NumNFTCopiesBurned)
+	}
+
+	// Roll all successful txns through connect and disconnect loops to make sure nothing breaks.
+	_rollBackTestMetaTxnsAndFlush(testMeta)
+	_applyTestMetaTxnsToMempool(testMeta)
+	_applyTestMetaTxnsToViewAndFlush(testMeta)
+	_disconnectTestMetaTxnsFromViewAndFlush(testMeta)
+	_connectBlockThenDisconnectBlockAndFlush(testMeta)
```
Comment (diamondhands0) [inline lib/block_view_test.go:19711]: Man, what a nice battery of tests we're doing here.

File: lib/constants.go [hunk 11]
```diff
@@ -118,6 +118,10 @@ var (
 	// rather than in creator coin.
 	// Triggers: 3pm PT on 8/16/2021
 	BitCloutDiamondsBlockHeight = uint32(52112)
+
+	// NFTTransfersBlockHeight defines the height at which NFT transfer txns, accept NFT
+	// transfer txns, and NFT burn txns will be accepted.
+	NFTTransferOrBurnBlockHeight = uint32(100000)
```
Comment (diamondhands0) [inline lib/constants.go:124]: Let's merge with this set to the really high number then do a follow-up PR with it set to whenever we think people can upgrade by. I'm thinking Monday or Tuesday of next week could be good if we merge it quickly.

File: lib/block_view.go [hunk 12]
```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
...
+		OwnerPKID:                  receiverPKID.PKID,
+		NFTPostHash:                prevNFTEntry.NFTPostHash,
+		SerialNumber:               prevNFTEntry.SerialNumber,
+		IsForSale:                  prevNFTEntry.IsForSale,
+		MinBidAmountNanos:          prevNFTEntry.MinBidAmountNanos,
+		UnlockableText:             txMeta.UnlockableText,
+		LastAcceptedBidAmountNanos: prevNFTEntry.LastAcceptedBidAmountNanos,
+		IsPending:                  true,
+	}
+	bav._setNFTEntryMappings(newNFTEntry)
```
Comment (diamondhands0) [inline lib/block_view.go:7267]: So I think our convention in general is to call _deleteNFTEntryMappings(prevNFTEntry) *and then* call _setNFTEntryMappings(newNFTEntry). Like so:

```
bav._deleteNFTEntryMappings(prevNFTEntry)
bav._setNFTEntryMappings(newNFTEntry)
```

I think it works as-is, but doing it this way future-proofs us in the event we create additional indexes in the view. For example, if we at some point created a mapping that used OwnerPKID as a key, then we would need to call _delete first to remove the old mapping.

I would do this in all the _connect *and* _disconnect functions btw.

Comment (diamondhands0) [inline lib/block_view.go:7267]: So I think our convention in general is to call _deleteNFTEntryMappings(prevNFTEntry) *and then* call _setNFTEntryMappings(newNFTEntry). Like so:

```
bav._deleteNFTEntryMappings(prevNFTEntry)
bav._setNFTEntryMappings(newNFTEntry)
```

I think it works as-is, but doing it this way future-proofs us in the event we create additional indexes in the view. For example, if we at some point created a mapping that used OwnerPKID as a key, then we would need to call _delete first to remove the old mapping.

I would do this in all the _connect *and* _disconnect functions btw.

File: lib/block_view.go (hunk 1 above)
Comment (redpartyhat) [reply lib/block_view.go:6507]: I don't see any harm in making things more strict.  Added.

File: lib/block_view.go (hunk 3 above)
Comment (redpartyhat) [reply lib/block_view.go:7256]: No reason, just copied it from _connectCreateNFT.  Changed.

File: lib/block_view.go (hunk 5 above)
Comment (redpartyhat) [reply lib/block_view.go:7351]: I like it. Cleaner.

Comment (redpartyhat) [reply lib/block_view.go:7351]: Appears to be a duplicate comment.  Resolving.

**Comment by redpartyhat** [comment]
All comments have been addressed.  I also added an "rph/nft-transfers" branch for backend, frontend, and identity. These branches were used to test the changes locally and validate the new transaction types worked as expected.

**Review by diamondhands0 — APPROVED** [review]

**Review by superzordon — APPROVED** [review]

**Pull request merged**